arraycontext/container/dataclass.py

```diff
@@ -96,12 +106,11 @@ def is_array_field(f: Field, field_type: type) -> bool:
                         f"Field '{f.name}' union contains non-array container "
                         "arguments. All arguments must be array containers.")
 
-        if isinstance(field_type, str):
-            raise TypeError(
-                f"String annotation on field '{f.name}' not supported. "
-                "(this may be due to 'from __future__ import annotations')")
-
         if __debug__:
+            # NOTE: this should never happen due to using `inspect.get_annotations`
+            if isinstance(field_type, str):
+                raise TypeError(f"String annotation on field '{f.name}' not supported.")
```

Comment:
```suggestion
                raise AssertionError(f"Unexpected string annotation on field '{f.name}'.")
```

Reply:
Went with a simple assert in https://github.com/inducer/arraycontext/compare/c44687e7444bd03b4b09c38bc19ea4125727325c..f705a8ad0bedc57175942f91dc07f08bdcc48557.